File: event.go
package apptrace

import (
	"fmt"
	"reflect"
	"time"
)

// An Event is a record of the occurrence of something.
type Event interface {
	Schema() string
}

// EventMarshaler is the interface implemented by an event that can
// marshal a representation of itself into annotations.
//
// TODO(sqs): implement this in MarshalEvent
type EventMarshaler interface {
	MarshalEvent() ([]*Annotation, error)
}

// EventUnmarshaler is the interface implemented by an event that can
// unmarshal an annotation representation of itself.
//
// TODO(sqs): implement this in UnmarshalEvent
type EventUnmarshaler interface {
	UnmarshalEvent([]*Annotation) error
}

const schemaPrefix = "_schema:"

// MarshalEvent marshals an event into annotations.
func MarshalEvent(e Event) (Annotations, error) {
	var as Annotations
	flattenValue("", reflect.ValueOf(e), func(k, v string) {
		as = append(as, Annotation{Key: k, Value: []byte(v)})
	})
	as = append(as, Annotation{Key: schemaPrefix + e.Schema()})
	return as, nil
}

// An EventSchemaUnmarshalError is when annotations are attempted to
// be unmarshaled into an event object that does not match any of the
// schemas in the annotations.
type EventSchemaUnmarshalError struct {
	Found  []string // schemas found in the annotations
	Target string   // schema of the target event
}

func (e *EventSchemaUnmarshalError) Error() string {
	return fmt.Sprintf("event: can't unmarshal annotations with schemas %v into event of schema %s", e.Found, e.Target)
}

// UnmarshalEvent unmarshals annotations into an event.
func UnmarshalEvent(as Annotations, e Event) error {
	aSchemas := as.schemas()
	schemaOK := false
	for _, s := range aSchemas {
		if s == e.Schema() {
			schemaOK = true
			break
		}
	}
	if !schemaOK {
		return &EventSchemaUnmarshalError{Found: aSchemas, Target: e.Schema()}
	}

	unflattenValue("", reflect.ValueOf(&e), reflect.TypeOf(&e), mapToKVs(as.StringMap()))
	return nil
}

// A spanName event sets a span's name.
type spanName string

func (spanName) Schema() string { return nameKey }

// Msg returns an Event that contains only a human-readable message.
func Msg(msg string) Event {
	return msgEvent{Msg: msg}
}

type msgEvent struct {
	Msg string
}

func (msgEvent) Schema() string { return "msg" }

// A TimestampedEvent is an Event with a timestamp.
type TimestampedEvent interface {
	Timestamp() time.Time
}

// Log returns an Event whose timestamp is the current time that
// contains only a human-readable message.
func Log(msg string) Event {
	return logEvent{Msg: msg, Time: time.Now()}
}

type logEvent struct {
	Msg  string
	Time time.Time
}

func (logEvent) Schema() string { return "log" }

func (e *logEvent) Timestamp() time.Time { return e.Time }

File: python/appdash/event.py
import strict_rfc3339
import time
from spanid import *

__schemaPrefix = "_schema:"

# timeString returns a appdash-compatable (RFC 3339 / UTC offset) time string
# for the given timestamp (float seconds since epoch).
def timeString(ts):
    return strict_rfc3339.timestamp_to_rfc3339_utcoffset(ts)

# MarshalEvent marshals an event into annotations.
def MarshalEvent(e):
    a = []
    for key, value in e.marshal().items():
        a.append(Annotation(key, value))
    a.append(Annotation(__schemaPrefix + e.schema, ""))
    return a

# SpanNameEvent is an event which sets a span's name.
class SpanNameEvent:
    schema = "name"

    # name is literally the span's name.
    name = ""

    def __init__(self, name):
        self.name = name

    # marshal returns a dictionary of this event's value by name.
    def marshal(self):
        return {"Name": self.name}

# MsgEvent is an event that contains only a human-readable message.
class MsgEvent:
    schema = "msg"

    # msg is literally the message string.
    msg = ""

    def __init__(self, msg):
        self.msg = msg

    # marshal returns a dictionary of this event's value by schema.
    def marshal(self):
        return {"Msg": self.msg}

# LogEvent is an event whose timestamp is the current time and contains the
# given human-readable log message.
class LogEvent:
    schema = "log"

    # msg is literally the message string.
    msg = ""

    # RFC3339-UTC timestamp of the event.
    time = ""

    def __init__(self, msg):
        self.msg = msg
        self.time = timeString(time.time())

    def marshal(self):
        return {"Msg": self.msg, "Time": self.time}

# SQLEvent is an SQL query event with send and receive times, as well as the
# actual SQL that was ran, and a optional tag.
class SQLEvent:
    schema = "SQL"

    # sql is literally the SQL query that was ran.
    sql = ""

    # tag is a optional user-created tag associated with the SQL event. 
    tag = ""

    # RFC3339-UTC timestamp of when the query was sent, and later a result received.
    clientSend = ""
    clientRecv = ""

    def __init__(self, sql, send, recv=None, tag=""):
        self.sql = sql
        self.tag = tag
        self.clientSend = timeString(send)

        # If user didn't specify a recv time, use right now.
        if recv:
            self.clientRecv = timeString(recv)
        else:
            self.clientRecv = timeString(time.time())

    def marshal(self):
        return {
            "SQL": self.sql,
            "Tag": self.tag,
            "ClientSend": self.clientSend,
            "ClientRecv": self.clientRecv,
        }

# TODO(slimsag): add HTTPEvent
